### PR TITLE
Redo Unstructured to have accessor methods

### PR DESCRIPTION
**Pull request by krousey**
Add accessor methods that implement pkg/api/unversioned.ObjectKind,
pkg/api/meta.Object, pkg/api/meta.Type and pkg/api/meta.List.

Removed the convenience fields since writing to them was not reflected
in serialized JSON.

### CODE REVIEW

**Comment by krousey** [comment]
cc @kubernetes/sig-api-machinery

File: pkg/runtime/register.go [hunk 1]
```diff
@@ -30,9 +30,10 @@ func (obj *TypeMeta) GroupVersionKind() *unversioned.GroupVersionKind {
 	return unversioned.FromAPIVersionAndKind(obj.APIVersion, obj.Kind)
 }
 
-func (obj *Unknown) GetObjectKind() unversioned.ObjectKind          { return &obj.TypeMeta }
-func (obj *Unstructured) GetObjectKind() unversioned.ObjectKind     { return &obj.TypeMeta }
-func (obj *UnstructuredList) GetObjectKind() unversioned.ObjectKind { return &obj.TypeMeta }
+func (obj *Unknown) GetObjectKind() unversioned.ObjectKind { return &obj.TypeMeta }
+
+func (obj *Unstructured) GetObjectKind() unversioned.ObjectKind     { return obj }
```
Comment (caesarxuchao) [inline pkg/runtime/register.go]: Shall we implement a GetObjectMeta() as well?

```
func (obj *Unknown) GetObjectMeta() unversioned.Object { return obj }
```

Comment (krousey) [inline pkg/runtime/register.go]: Why?

**Comment by krousey** [comment]
@caesarxuchao PTAL

**Comment by caesarxuchao** [comment]
Could you add a line to check if `Unstructured` implements the `Object` interface? You can add the line in interface.go, where the `Object` is defined, otherwise there is an import cycle.

**Comment by krousey** [comment]
I don't understand why. No other type makes that check, and nothing changed with regards to implementing `Object`. It still has `GetObjectKind`, just implemented differently.

**Comment by caesarxuchao** [comment]
I mean the meta.Object (https://github.com/kubernetes/kubernetes/blob/master/pkg/api/meta/interfaces.go#L39). Sorry I didn't realize there is also an runtime.Object. 
meta.Object interface has many methods, so I think adding a check will be useful. I got a bite when I added more methods to that interface and forgot to update Unstructured.

**Comment by krousey** [comment]
Ah. Done.

**Comment by caesarxuchao** [comment]
LGTM. Thanks.

File: pkg/runtime/types.go [hunk 2]
```diff
@@ -120,26 +125,259 @@ type Unknown struct {
...
-
-	// Name is populated from metadata (if present) upon deserialization
-	Name string
-
 	// Object is a JSON compatible map with string, float, int, []interface{}, or map[string]interface{}
 	// children.
 	Object map[string]interface{}
 }
 
+func getNestedField(obj map[string]interface{}, fields ...string) interface{} {
```
Comment (smarterclayton) [inline pkg/runtime/types.go]: This is more of a follow up, but I'm pretty sure we have a library somewhere in our godeps that does this.  Just not coming to mind off the top of my head.  No change necessary here, but if we extend this we should switch.

**Pull request merged**